src/view/wlr-surface-node.cpp

```diff
@@ -205,7 +205,7 @@ wf::touch_interaction_t& wf::scene::wlr_surface_node_t::touch_interaction()
 
 void wf::scene::wlr_surface_node_t::send_frame_done(bool delay_until_vblank)
 {
-    if (!surface)
+    if (!wf::get_core().session->active || !surface)
```

Comment:
I don't think this is the correct place for such a check. Frame done events are sent in some specific edge-cases (pending transactions, surface not visible on any output but trying to get mapped, etc.) and when the output emits a `frame_done_signal`. In my estimation, the rest of your changes should be enough to fix the issue, if not, then probably we need a guard around the emission of `frame_done_signal` in render-manager.cpp